homeassistant/components/airvisual/__init__.py

```diff
@@ -1 +1,220 @@
+    )
+)
+
+CLOUD_API_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_GEOGRAPHIES, default=[]): vol.All(
+            cv.ensure_list, [GEOGRAPHY_SCHEMA]
+        ),
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): vol.All(
+            cv.time_period, lambda value: value.total_seconds()
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: CLOUD_API_SCHEMA}, extra=vol.ALLOW_EXTRA,)
+
+
+@callback
+def async_get_geography_id(geography_dict):
+    """Generate a unique ID from a geography_dict dict."""
+    if CONF_CITY in geography_dict:
+        return ",".join(
+            (
+                geography_dict[CONF_CITY],
+                geography_dict[CONF_STATE],
+                geography_dict[CONF_COUNTRY],
+            )
+        )
+    return ",".join(
+        (str(geography_dict[CONF_LATITUDE]), str(geography_dict[CONF_LONGITUDE]))
+    )
+
+
+async def async_setup(hass, config):
+    """Set up the AirVisual component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    if conf[CONF_API_KEY] in configured_instances(hass):
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up AirVisual as config entry."""
+    websession = aiohttp_client.async_get_clientsession(hass)
+
+    hass.data[DOMAIN][DATA_CLIENT][config_entry.entry_id] = AirVisualData(
+        hass,
+        Client(websession, api_key=config_entry.data[CONF_API_KEY]),
+        config_entry.data[CONF_GEOGRAPHIES],
+        config_entry.options.get(CONF_SHOW_ON_MAP, config_entry.data[CONF_SHOW_ON_MAP]),
+    )
+
+    try:
+        await hass.data[DOMAIN][DATA_CLIENT][config_entry.entry_id].async_update()
+    except InvalidKeyError:
+        _LOGGER.error("Invalid API key provided")
+        raise ConfigEntryNotReady
+
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(config_entry, "sensor")
+    )
+
+    async def refresh(event_time):
+        """Refresh data from AirVisual."""
+        await hass.data[DOMAIN][DATA_CLIENT][config_entry.entry_id].async_update()
+
+    hass.data[DOMAIN][DATA_LISTENER][config_entry.entry_id] = async_track_time_interval(
+        hass, refresh, timedelta(seconds=config_entry.data[CONF_SCAN_INTERVAL])
+    )
+
+    config_entry.add_update_listener(async_update_entry)
+
+    return True
+
+
+async def async_update_entry(hass, config_entry):
+    """Process an options update to the config entry."""
+    hass.data[DOMAIN][DATA_CLIENT][
+        config_entry.entry_id
+    ].show_on_map = config_entry.options[CONF_SHOW_ON_MAP]
```

Comment:
This seems backwards

Reply:
Not sure what you mean, but I like your idea of simplifying things by putting options in a different PR, so I’ll approach this then.